src/server/job_recov_db.c

```diff
@@ -545,14 +545,18 @@ job_recov_db(char *jid)
 
 	pj = job_recov_db_spl(&dbjob);
 	if (!pj)
-		goto db_err;
+		goto db_err_reset;
 
 	if (pbs_db_end_trx(conn, PBS_DB_COMMIT) != 0)
-		goto db_err;
+		goto db_err_reset;
 
 	pbs_db_reset_obj(&obj);
 
 	return (pj);
+
+db_err_reset:
```

Comment:
Far i recall, we can call pbs_db_reset_obj() directly - the function checks whether to free or not. So perhaps we can it anyways db_err label itself, instead of making another label db_err_reset. Please check that.

Reply:
Hey @subhasisb I had the same plan initially but after analyzing the db interface code I saw conditional jump on uninitialized stack memory in `free_db_attr_list()` if the failure happens before a call to `convert_array_to_db_attr_list()`, hence chose the new label. I spent some effort and tried to simulate this flow and as suspected got below result using valgrind
```
==31371== Conditional jump or move depends on uninitialised value(s)
==31371==    at 0x53E77A: free_db_attr_list (db_postgres_attr.c:222)
==31371==    by 0x53BFAF: pg_db_reset_que (db_postgres_que.c:372)
==31371==    by 0x535F84: pbs_db_reset_obj (db_postgres_impl.c:903)
==31371==    by 0x4844C5: que_recov_db (queue_recov_db.c:243)
==31371==    by 0x47A431: pbsd_init (pbsd_init.c:889)
==31371==    by 0x47F5E1: main (pbsd_main.c:1691)
==31371==  Uninitialised value was created by a stack allocation
==31371==    at 0x4843F9: que_recov_db (queue_recov_db.c:212)
```

I agree, we can remove the label `db_err_reset`, if we agree on initializing the stack memory `pbs_db_que_info_t	dbque;` to zero in `que_recov_db()` and similar `*_recov_db()` functions?

Reply:
In addition to the Conditional jump on uninitialised values, I'm also seeing many invalid reads and invalid free's
I'm attaching this valgrind log below
[valgrind_simulated_error.log](https://github.com/PBSPro/pbspro/files/3828701/valgrind_simulated_error.log)

Reply:
Yes, so the attribute array is uninitialized before read, we can actually initialize it pretty easily. 
If we could initialize the dbque attribute list thus:
	dbque.attr_list.attributes = NULL;
	dbque.attr_list.attr_count = 0;

So, your set of lines at the top of que_recov_db() (similarly for jobs and other objects as well) would then be:
	obj.pbs_db_obj_type = PBS_DB_QUEUE;
	obj.pbs_db_un.pbs_db_que = &dbque;
	dbque.attr_list.attributes = NULL;
	dbque.attr_list.attr_count = 0;


The alternative is to do it inside pg_db_load_que() (similarly for jobs and other objects as well):
	pq->attr_list.attr_count = 0;
	pq->attr_list.attributes = NULL;

This would be better than adding another goto label.

Reply:
> If we could initialize the dbque attribute list thus:
> dbque.attr_list.attributes = NULL;
> dbque.attr_list.attr_count = 0;
> 

Or how about this one line init of full struct ? I dont feel the object is too big for it to have a penalty.
`pbs_db_que_info_t dbque = {0};`

Reply:
Yah, that sounds fine too. Yes, it is not too big to have a penalty.

Reply:
Fixed in https://github.com/PBSPro/pbspro/pull/1394/commits/019f046bf1570332e453e9d6451c79abfe7d5bb7

tests repeated and attached in the header!